File: algorithms/260. Single Number III/solution.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        x = reduce(operator.xor, nums)
        # get last set bit, which is a diff bit for n1 and n2
        x &= -x
        n1 = n2 = 0
        for n in nums:
            if n & x:
                n1 ^= n
            else:
                n2 ^= n
        return [n1, n2]

File: algorithms/260. Single Number III/solution.cc
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int x = 0;
        for ( int n : nums ) {
            x ^= n;
        }
        // get last set bit, which is a diff bit for n1 and n2
        x &= -x;
        vector<int> single(2);
        for ( int n : nums ) {
            if ( n & x ) {
                single[0] ^= n;
            } else {
                single[1] ^= n;
            }
        }
        return single;
    }
};
